src/main/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTask.java

```diff
@@ -146,48 +142,129 @@ private static void checkRateLimit(final MongoSinkTopicConfig config)
     }
   }
 
-  private static void handleMongoException(
-      final MongoSinkTopicConfig config,
-      final List<WriteModel<BsonDocument>> writeModels,
-      final MongoException e) {
-    if (config.logErrors()) {
-      LOGGER.error("Error on mongodb operation", e);
-      if (e instanceof MongoBulkWriteException) {
-        LOGGER.error("Mongodb bulk write (partially) failed", e);
-        LOGGER.error("WriteResult: {}", ((MongoBulkWriteException) e).getWriteResult());
-        LOGGER.error(
-            "WriteErrors: {}",
-            generateWriteErrors(((MongoBulkWriteException) e).getWriteErrors(), writeModels));
-        LOGGER.error("WriteConcernError: {}", ((MongoBulkWriteException) e).getWriteConcernError());
+  private void handleTolerableWriteException(
+      final List<MongoProcessedSinkRecordData> batch,
+      final RuntimeException e,
+      final boolean logErrors,
+      final boolean tolerateErrors) {
+    List<SinkRecord> records =
+        batch.stream()
+            .map(MongoProcessedSinkRecordData::getSinkRecord)
+            .collect(Collectors.toList());
+    if (e instanceof MongoBulkWriteException) {
+      AnalyzedBatchFailedWithBulkWriteException analyzedBatch =
+          new AnalyzedBatchFailedWithBulkWriteException(records, (MongoBulkWriteException) e);
+      if (logErrors) {
+        analyzedBatch.log();
+      }
+      if (tolerateErrors) {
+        analyzedBatch.report();
+      } else {
+        throw new DataException(e);
+      }
+    } else {
+      if (logErrors) {
+        log(records, e);
+      }
+      if (!tolerateErrors) {
+        throw new DataException(e);
       }
-    }
-    if (!config.tolerateErrors()) {
-      throw new DataException("Failed to write mongodb documents", e);
     }
   }
 
-  private static String generateWriteErrors(
-      final List<BulkWriteError> bulkWriteErrorList,
-      final List<WriteModel<BsonDocument>> writeModels) {
-    List<String> errorString = new ArrayList<>();
-    for (final BulkWriteError bulkWriteError : bulkWriteErrorList) {
-      if (bulkWriteError.getIndex() < writeModels.size()) {
-        errorString.add(
-            "BulkWriteError{"
-                + "writeModel="
-                + writeModels.get(bulkWriteError.getIndex())
-                + ", code="
-                + bulkWriteError.getCode()
-                + ", message='"
-                + bulkWriteError.getMessage()
-                + '\''
-                + ", details="
-                + bulkWriteError.getDetails()
-                + '}');
+  private class AnalyzedBatchFailedWithBulkWriteException {
```

Comment:
Consider moving into its own file - only uses the error reporter which can be passed in and the static logger.

Reply:
Perhaps add a comment as to the purpose of this class

Reply:
Done, moved to the `com.mongodb.kafka.connect.sink.dlq` package together with DQL-related exceptions.